File: 04_schedule/problems/00_schedule_dynamic/schedule_dynamic.f90
module work_mod
contains
  ! 仕事量が引数 k に比例するダミー計算 (k に比例した回数だけ加算する)
  function work(k) result(s)
    integer(8), intent(in) :: k
    real(8) :: s
    integer(8) :: j
    s = 0.0d0
    do j = 0, k - 1
       s = s + 1.0d0 / (1.0d0 + j)
    end do
  end function work
end module work_mod

program schedule_dynamic
  use omp_lib
  use work_mod
  integer, parameter :: n = 2000
  integer :: i
  real(8) :: total
  total = 0.0d0
  ! TODO: 下のループを並列化し, 仕事量が i に比例して不均一なので schedule(dynamic) で負荷を均せ.
  do i = 0, n - 1
     ! 繰り返し i の仕事量は i に比例して重くなる (アンバランス)
     total = total + work(int(i, 8) * 100000_8)
  end do
  ! TODO: 上で始めた parallel do 領域を閉じる (!$omp end parallel do).
  print "(a,f0.6)", "total = ", total
end program schedule_dynamic

File: 04_schedule/problems/00_schedule_dynamic/schedule_dynamic.cpp
#include <cstdio>
#include <omp.h>

/* 仕事量が引数 k に比例するダミー計算 (k に比例した回数だけ加算する) */
double work(long k) {
  double s = 0.0;
  for (long j = 0; j < k; j++) {
    s += 1.0 / (1.0 + j);
  }
  return s;
}

int main() {
  const int n = 2000;
  double total = 0.0;
  // TODO: 下のループを並列化し, 仕事量が i に比例して不均一なので schedule(dynamic) で負荷を均せ.
  for (int i = 0; i < n; i++) {
    /* 繰り返し i の仕事量は i に比例して重くなる (アンバランス) */
    total += work((long)i * 100000L);
  }
  printf("total = %f\n", total);
  return 0;
}
